sdk/metric/reader.go

```diff
@@ -0,0 +1,92 @@
+// Copyright The OpenTelemetry Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package metric
+
+import (
+	"context"
+
+	"go.opentelemetry.io/otel/sdk/metric/export"
+	"go.opentelemetry.io/otel/sdk/metric/view"
+)
+
+// Reader is the interface used between the SDK and an
+// exporter.  Control flow is bi-directional through the
+// Reader, since the SDK initiates ForceFlush and Shutdown
+// while the initiates collection.  The Register() method here
+// informs the Reader that it can begin reading, signaling the
+// start of bi-directional control flow.
+//
+// Typically, push-based exporters that are periodic will
+// implement PeroidicExporter themselves and construct a
+// PeriodicReader to satisfy this interface.
+//
+// Pull-based exporters will typically implement Register
+// themselves, since they read on demand.
+type Reader interface {
+	// Register is called when the SDK is fully
+	// configured.  The Producer passed allows the
+	// Reader to begin collecting metrics using its
+	// Produce() method.
+	register(producer)
```

Comment:
I propose exporting `Register()`, `Producer`, and `Produce()`.  Why not allow users to provide their own?

By exporting the Producer() and Produce() method, any exporter can implement custom Reader logic, for example an alternative periodic reader that could be paired with existing PushExporters.

Reply:
Isolating the register method helps us evolve the API as it gets more use. It does not prevent us from exporting it at a later date.

For alternate implementation they can wrap the manual producer with their own periodic logic given it has its own `Collect` method.

Reply:
> Isolating the register method helps us evolve the API as it gets more use. It does not prevent us from exporting it at a later date.

Haven't the ManualReader and PeriodicReader implementations demonstrated enough use? I can't see what incompatible changes would be allowed at this point, given where the specification is.  It also seems extraneous to force the ManualReader to make decisions about synchronization and allow configuring alternative Shutdown and ForceFlush behavior, when there are a myriad of behaviors they might want and can easily implement whatever they want if we expose the Reader/Producer/Register methods.